### PR TITLE
Fix `contrib/build_sysimg.jl`

### PR DESCRIPTION
**Pull request by staticfloat**
The `contrib/build_sysimg.jl` script had bitrotted a bit with the 0.7 changes.  This PR also incorporates @RalphAS's changes to address https://github.com/JuliaLang/julia/issues/27201 and https://github.com/JuliaLang/julia/issues/27451.

Marking this as WIP until testing is done on all major OS's:

- [x] Linux
- [x] OSX
- [x] FreeBSD
- [x] Windows

@ararslan could I ask you to test this out on FreeBSD?  I create the "user image" file:

```julia
# /tmp/usrimg.jl
function foobar()
    println("It worked!")
end
```

Then I compile it into a custom sysimg:
```
$ ./julia --color=yes ./contrib/build_sysimg.jl /tmp/sysimg x86-64 /tmp/usrimg.jl
```

Finally, I test it:
```
$ ./julia -J /tmp/sysimg.so -e 'foobar()'
It worked!
```

### CODE REVIEW

**Comment by ViralBShah** [comment]
I believe PackageCompiler.jl is the more general way to do this.

**Comment by ararslan** [comment]
If this script is no longer the recommended way to accomplish this, can we simply delete the script and tell people to use PackageCompiler?

File: contrib/build_sysimg.jl [hunk 1]
```diff
@@ -164,18 +162,14 @@ function link_sysimg(sysimg_path=nothing, cc=find_system_compiler(), debug=false
...
+    @info("Linking sys.$(Libdl.dlext) with $cc_cmd")
     # Windows has difficulties overwriting a file in use so we first link to a temp file
-    if Sys.iswindows() && isfile(sysimg_file)
-        if success(pipeline(`$cc $FLAGS -o $sysimg_path.tmp $sysimg_path.o`; stdout=stdout, stderr=stderr))
-            mv(sysimg_file, "$sysimg_file.old"; force=true)
-            mv("$sysimg_path.tmp", sysimg_file; force=true)
-        end
-    else
-        run(`$cc $FLAGS -o $sysimg_file $sysimg_path.o`)
+    if success(pipeline(cc_cmd; stdout=stdout, stderr=stderr))
```
Comment (ararslan) [inline contrib/build_sysimg.jl]: One of the `mv` calls here is causing an `ENOENT` on FreeBSD.

**Comment by ararslan** [comment]
Looks good on FreeBSD now 👍

**Comment by ararslan** [comment]
Getting this on macOS:
```
[ Info: Linking sys.dylib with `cc -L/Users/alex/Projects/julia/usr/lib -shared -ljulia -o /Users/alex/stuff/sysimg.tmp -Wl,--whole-archive /Users/alex/stuff/sysimg.o -Wl,--no-whole-archive`
ld: unknown option: --whole-archive
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
I think something like this should fix it:
```diff
--- a/contrib/build_sysimg.jl
+++ b/contrib/build_sysimg.jl
@@ -161,8 +161,16 @@ function link_sysimg(sysimg_path=nothing, cc=find_system_compiler(), debug=false
         push!(FLAGS, "-lssp")
     end

+    if Sys.isapple()
+        whole_archive = "-Xlinker -all_load"
+        no_whole_archive = ""
+    else
+        whole_archive = "-Wl,--whole-archive"
+        no_whole_archive = "-Wl,--no-whole-archive"
+    end
+
     sysimg_file = "$sysimg_path.$(Libdl.dlext)"
-    cc_cmd = `$cc $FLAGS -o $sysimg_path.tmp -Wl,--whole-archive $sysimg_path.o -Wl,--no-whole-archive`
+    cc_cmd = `$cc $FLAGS -o $sysimg_path.tmp $whole_archive $sysimg_path.o $no_whole_archive`
     @info("Linking sys.$(Libdl.dlext) with $cc_cmd")
     # Windows has difficulties overwriting a file in use so we first link to a temp file
     if success(pipeline(cc_cmd; stdout=stdout, stderr=stderr))
```

**Comment by tknopp** [comment]
I always wondered why this needs to be a script and is not a regular function within julia. But I tend to agree with @ararslan that if `PackageCompiler.jl` does this and well tested, it might be better having the functionality only once.

**Comment by dgleich** [comment]
Just a quick note on this. In the past I compiled custom sysimages to get things like fma operations to use native instructions when they were supported. First, this no longer seems to be needed. But second, the way I did this was: `bin/julia share/julia/build_sysimg.jl --force` which now seems to be a good way to ruin your julia install :)

I get the 
> ERROR: System image file failed consistency check: maybe opened the wrong version?
when trying to use the new one. 

This is on linux with both julia 0.7.0 and julia 1.0.0 (after I modified the script to use @info and using Libdl)

**Comment by Gnimuc** [comment]
I'm on macOS, also got this error "ERROR: System image file failed consistency check: maybe opened the wrong version?".

**Comment by SimonDanisch** [comment]
If someone more familiar with correctly linking + outputting the shared libraries etc helps to keep PackageCompiler up to date, I'd very much welcome to just use PackageCompiler instead ;)

Otherwise it'd be nice to keep this script as a reference implementation!
It's always a bit of an adventure for me to figure out what exact flags && commands need to go into PackageCompiler after a Julia compiler change...

**Comment by staticfloat** [comment]
I think we should just remove `contrib/build_sysimg.jl` and tell users to use `PackageCompiler.jl` instead.  All in favor?

**Comment by ViralBShah** [comment]
Please delete.

**Comment by mattcbro** [comment]
What?  PackageCompiler.jl doesn't build  post 1.0.  I'm getting Libdl errors just trying to run default_sysimg_path() .   I'm on Linux Mint 19.  Are there any work arounds?

**Comment by staticfloat** [comment]
I suggest you open an issue against `PackageCompiler.jl` if it isn't working for you; it should definitely work on 1.0.

**Pull request closed without merging**